### PR TITLE
feat(sandbox,lib,demo): modal in portal

### PR DESCRIPTION
**Pull request by johannes-lindgren**
## What?

There is an issue that makes various elements on the screen overlap the field plugin modal:

image

This happens because the field plugin is not using Teleport, but instead uses dynamic styling on its wrapping element to go from `position: relative` to `position: absolute`. But the iframe (and its wrapping div element) might very well not have the highest priority in its stacking context. The way this is normally solved in dialogs is to use `<Teleport>` to mount the element outside the stacking context, (somewhere next to the application root element).

However, this is not possible to implement without some kind of a breaking change: without the `Teleport`, the iframe element never remounts, which means that the plugin's application state is preserved between. transitions between modal and non-modal modes. But if we conditionally render it inside a `<Teleport>`, the iframe element is going to remount, which reset the plugin's application state. Some plugins persist their entire app state in the content—but not all, which is why this would be a breaking change.

Rather than making the breaking change in Storyfront, we are introducing an option `enablePortalModal` in `@storyblok/field-plugin` which is sent to Storyfront on initialization. This means that the clients will have to manually enable the flag to access this bug ifx. _This means that the bug with the overlapping element will persist for all existing field plugins until the authors upgrade and enable the new option._

### Changes

The solution is to include a new option `enablePortalModal` in the field-plugin SDK: when enabled, and when the field plugin loads, it will include a property `enablePortalModal: true` in the message to initialize. When set to `true`, the modal will be done with a `<Teleport>`, otherwise in the old way.

This value is then combined with the `isModal` state (which can be changed by the field plugin or the editor) into a derived value `modalState` that is in either of the following states: `'non-modal' | 'modal-with-portal' | 'modal-without-portal'`.

The option `enablePortalModal` is set to `true` in all templates.

## How to test?

#420 makes it more difficult to test. The sandbox preview won't work because it's not a whitelisted host. So you need to serve the sandbox and demo apps in local development, and update `packages/demo/src/components/FieldPluginDemo.tsx`:

```ts
  const { type, data, actions } = useFieldPlugin({
    validateContent,
    // Storyfront in local development mode
    // targetOrigin: 'http://localhost:3300',
    // Sandbox in local development mode:
    targetOrigin: 'http://localhost:7070',
  })
``` 

### Sandbox

In local development mode, you need change the demo application's `targetOrigib`:

`packages/demo/src/components/FieldPluginDemo.tsx`

```tsx
export const FieldPluginDemo: FunctionComponent = () => {
  const { type, data, actions } = useFieldPlugin({
    validateContent,
    targetOrigin: 'http://localhost:7070',
  })
```

Then either test with a plugin on the previous versions, or just comment out this line:

`packages/field-plugin/src/messaging/pluginMessage/pluginToContainerMessage/PluginLoadedMessage.ts`:

The line `  enablePortalModal: true,` in

```ts
export const pluginLoadedMessage = (
  options: Pick<PluginLoadedMessage, 'uid' | 'callbackId'>,
): PluginLoadedMessage => ({
  action: 'plugin-changed',
  event: 'loaded',
  fullHeight: true,
  subscribeState: true,
  enablePortalModal: true,
  ...options,
})
```

### CODE REVIEW

**Comment by eunjae-lee** [comment]
Is it worth making a break change and bump the major version? I wouldn't do that until we have real major improvements. It can give false impression to users.

**Comment by johannes-lindgren** [comment]
> Is it worth making a break change and bump the major version? I wouldn't do that until we have real major improvements. It can give false impression to users.

I was contemplating whether we should pass `enablePortalModal` as an argument to `createFieldPlugin`. That would benefit all plugins that gets created from this point onward because we could set it as default in the starters. But the drawback with that approach is that we don't force older plugins to deal with the issue.

**Comment by johannes-lindgren** [comment]
What do you think, @demetriusfeijoo, @Dawntraoz, @BibiSebi, @eunjae-lee?

**Comment by demetriusfeijoo** [comment]
Hey @johannes-lindgren, sorry for the late reply.

I think we could go as you said, but just release a minor version and update our starters/templates to include `enablePortalModal: true` by default, so most of the new plugins created based on them are already covered.

Then, In the future, if a new feature introduces an unavoidable breaking change, we can enjoy the opportunity and make the enablePortalModal prop always true by default in the SDK. 🤔

**Comment by johannes-lindgren** [comment]
With a non-breaking change, how do we push users to upgrade? If the release is minor, users are going to increase the version to the latest and call it a day.

**Comment by BibiSebi** [comment]
As it is not a breaking change I would stick with minor. What we can think about is to share this change in our discord. Or even think about opening up discussions (https://github.com/storyblok/field-plugin/discussions/landing) to announce these changes. I would not force them. If they encounter the issue, they will search for answers in our docs/community channels anyway).

**Comment by johannes-lindgren** [comment]
> As it is not a breaking change I would stick with minor. What we can think about is to share this change in our discord. Or even think about opening up discussions (https://github.com/storyblok/field-plugin/discussions/landing) to announce these changes. I would not force them. If they encounter the issue, they will search for answers in our docs/community channels anyway).

I think the users are more likely to catch this bug, and they will report it via helpdesk.

**Comment by johannes-lindgren** [comment]
@demetriusfeijoo, @BibiSebi, @Dawntraoz, I added an option to `createFieldPlugin`.

@BibiSebi, is it okay to update the templates like this in this PR? They are fetched from GitHub right? It it fetched with a tag? I mean, if we upgrade the core library, do we need to release a new version of the CLI as well?

**Comment by demetriusfeijoo** [comment]
Hey @johannes-lindgren 🙌 

I started reviewing it but didn't manage to finish it yet.
I'll focus on this and the one on the Storyfront side tomorrow.

> These are the ones I can think of:
Ecommerce
Bynder
Table (used by customers)
Annotated image (not sure how important)
Did I miss some?

Oh, I didn't know about `Table` and `Annotated image`. 
From the top of my mind, I'd just include the [Picker Starter](https://github.com/storyblok/field-type-examples/tree/main/picker-starter) on that list.

File: packages/demo/src/components/FieldPluginDemo.tsx [hunk 1]
```diff
@@ -19,6 +19,8 @@ export type PluginComponent = FunctionComponent<{
 export const FieldPluginDemo: FunctionComponent = () => {
   const { type, data, actions } = useFieldPlugin({
     validateContent,
+    targetOrigin: 'http://localhost:7070',
```
Comment (johannes-lindgren) [inline packages/demo/src/components/FieldPluginDemo.tsx:22]:
```suggestion
```

**Comment by BibiSebi** [comment]
After reviewing the changes on Storyfront, I see that major version indeed might be a better solution here. Let's go with that but let's discuss before hand if we want to somehow add a changelog or a new section to the readme explaining the breaking change? 

Also we need to update our docs with the new option. 

CC: @johannes-lindgren @demetriusfeijoo @Dawntraoz

**Comment by johannes-lindgren** [comment]
@BibiSebi, @demetriusfeijoo, @Dawntraoz; should I revert my recent changes that introduced the option? Or should I simply merge this as-is, and then follow up with a separate PR that makes this option the default?

**Comment by johannes-lindgren** [comment]
I know that Bynder doesn't have continuous deployment set up, but do the e-commerce plugins have that?

Also, have we upgraded all the old e-commerce plugins yet?

**Comment by Dawntraoz** [comment]
@johannes-lindgren @BibiSebi @demetriusfeijoo After reading the discussions, these are my 2 cents:

Johannes, can you include the option in the README.md so it is already documented? In that case, I would then merge as-is, release a major version for `field-plugin`, and then the same for `field-plugin-cli` afterwards, as per the release guidelines: https://github.com/storyblok/field-plugin/blob/main/CONTRIBUTING.md#release.

For the CD on e-commerce plugins, maybe @demetriusfeijoo can answer surely, but as far as I can see in the repo field-plugin e-commerce, all of them have, but they're still some missing in that repo 🤔

**Review by demetriusfeijoo — COMMENTED** [review]
Hey @johannes-lindgren 🙌 

> @BibiSebi, @demetriusfeijoo, @Dawntraoz; should I revert my recent changes that introduced the option? Or should I simply merge this as-is, and then follow up with a separate PR that makes this option the default?

I think it could be done in a separate PR 👍 

> I know that Bynder doesn't have continuous deployment set up, but do the e-commerce plugins have that?

All the e-commerce integrations in the main branch have continuous deployment 

> Also, have we upgraded all the old e-commerce plugins yet?

Not yet 😢 


As Alba mentioned there was one TS error related to the new added option:

image

I have left a comment about the tests which I think may be wrong but overall, the change looks good to me. 👏 👏

**Comment by johannes-lindgren** [comment]
> @johannes-lindgren @BibiSebi @demetriusfeijoo After reading the discussions, these are my 2 cents:
> 
> Johannes, can you include the option in the README.md so it is already documented? In that case, I would then merge as-is, release a major version for `field-plugin`, and then the same for `field-plugin-cli` afterwards, as per the release guidelines: https://github.com/storyblok/field-plugin/blob/main/CONTRIBUTING.md#release.
> 
> For the CD on e-commerce plugins, maybe @demetriusfeijoo can answer surely, but as far as I can see in the repo field-plugin e-commerce, all of them have, but they're still some missing in that repo 🤔

Ok so I should revert my recent changes? After the initial feedback, I did add it as an option, and set the default to be disabled.

File: packages/field-plugin/src/messaging/pluginMessage/pluginToContainerMessage/PluginLoadedMessage.test.ts [hunk 2]
```diff
@@ -76,6 +76,96 @@ describe('PluginLoadedMessage', () => {
...
+          }),
+        ).toEqual(false)
+        expect(
+          isPluginLoadedMessage({
+            ...stub,
+            subscribeState: null,
+          }),
+        ).toEqual(false)
+      })
+    })
```
[the remark this answers is not in the document]
Comment (johannes-lindgren) [reply packages/field-plugin/src/messaging/pluginMessage/pluginToContainerMessage/PluginLoadedMessage.test.ts:168]: Yes, I don't know what heppened here, heh

[the remark this answers is not in the document]
Comment (johannes-lindgren) [reply packages/field-plugin/src/messaging/pluginMessage/pluginToContainerMessage/PluginLoadedMessage.test.ts]: b0d2b9aebdd0076c8cc4e2d98174060e7506d563

**Comment by BibiSebi** [comment]
Update: After a conversation with @johannes-lindgren we decided to go for minor version update and first see how users react to it and just provide docs. In case we get complaints and support tickets we can always upgrade to major.

CC: @Dawntraoz @demetriusfeijoo

**Review by BibiSebi — APPROVED** [review]

**Comment by Dawntraoz** [comment]
> > @johannes-lindgren @BibiSebi @demetriusfeijoo After reading the discussions, these are my 2 cents:
> > Johannes, can you include the option in the README.md so it is already documented? In that case, I would then merge as-is, release a major version for `field-plugin`, and then the same for `field-plugin-cli` afterwards, as per the release guidelines: https://github.com/storyblok/field-plugin/blob/main/CONTRIBUTING.md#release.
> > For the CD on e-commerce plugins, maybe @demetriusfeijoo can answer surely, but as far as I can see in the repo field-plugin e-commerce, all of them have, but they're still some missing in that repo 🤔
> 
> @Dawntraoz, should I revert my recent changes? After the initial feedback, I did add it as an option, and set the default to be disabled, so that it wouldn't be a major version.
> 
> I moved the documentation to storyblok.com, so there's nothing in the readme file

If you disabled it and it is just the templates, then indeed, I agree, no need for a major version.

I mentioned the README because I saw @BibiSebi mentioned that it would be nice to include it in case of a breaking change, so it is quite clear for anyone looking for answers. But since it will be optional, and then released as a minor version, can you create a ticket to the docs team already to add the option in the main docs: https://www.storyblok.com/docs/plugins/field-plugins/introduction? 🙏 Thanks in advance!

**Review by Dawntraoz — APPROVED** [review]

**Review by gustavomelki — APPROVED** [review]

**Comment by johannes-lindgren** [comment]
> > > @johannes-lindgren @BibiSebi @demetriusfeijoo After reading the discussions, these are my 2 cents:
> > > Johannes, can you include the option in the README.md so it is already documented? In that case, I would then merge as-is, release a major version for `field-plugin`, and then the same for `field-plugin-cli` afterwards, as per the release guidelines: https://github.com/storyblok/field-plugin/blob/main/CONTRIBUTING.md#release.
> > > For the CD on e-commerce plugins, maybe @demetriusfeijoo can answer surely, but as far as I can see in the repo field-plugin e-commerce, all of them have, but they're still some missing in that repo 🤔
> > 
> > 
> > @Dawntraoz, should I revert my recent changes? After the initial feedback, I did add it as an option, and set the default to be disabled, so that it wouldn't be a major version.
> > I moved the documentation to storyblok.com, so there's nothing in the readme file
> 
> If you disabled it and it is just the templates, then indeed, I agree, no need for a major version.
> 
> I mentioned the README because I saw @BibiSebi mentioned that it would be nice to include it in case of a breaking change, so it is quite clear for anyone looking for answers. But since it will be optional, and then released as a minor version, can you create a ticket to the docs team already to add the option in the main docs: https://www.storyblok.com/docs/plugins/field-plugins/introduction? 🙏 Thanks in advance!

Thanks for the clarification. I agree the docs are essential. I will personally update the docs 🙂  (I wrote them: https://www.storyblok.com/docs/plugins/field-plugins/storyblok-field-plugin)

**Pull request merged**